concurrent_map.go

```diff
@@ -5,75 +5,114 @@ import (
 	"sync"
 )
 
-var SHARD_COUNT = 32
-
-// A "thread" safe map of type string:Anything.
-// To avoid lock bottlenecks this map is dived to several (SHARD_COUNT) map shards.
-type ConcurrentMap []*ConcurrentMapShared
-
-// A "thread" safe string to anything map.
-type ConcurrentMapShared struct {
-	items        map[string]interface{}
+// ShardCount represents the number of shards the map is divided into.
+var ShardCount = 32
+
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
+type ConcurrentMap []*ConcurrentMapShard
+
+// ConcurrentMapShared is a "thread" safe string to anything map.
+type ConcurrentMapShard struct {
+	maxSize uint32
+	idxAdd  uint32
+	idxLast uint32
+	size    uint32
+	mapKeys []string
+	items   map[string]interface{}
 	sync.RWMutex // Read Write mutex, guards access to internal map.
 }
 
-// Creates a new concurrent map.
-func New() ConcurrentMap {
-	m := make(ConcurrentMap, SHARD_COUNT)
-	for i := 0; i < SHARD_COUNT; i++ {
-		m[i] = &ConcurrentMapShared{items: make(map[string]interface{})}
+// New creates a new concurrent map.
+func New(maxSize uint32) ConcurrentMap {
+	m := make(ConcurrentMap, ShardCount)
+	shardSize := maxSize / uint32(ShardCount)
+	if maxSize%uint32(ShardCount) != 0 {
+		shardSize++
+	}
+	for i := 0; i < ShardCount; i++ {
+		m[i] = &ConcurrentMapShard{
+			maxSize: shardSize,
+			idxAdd:  0,
+			idxLast: 0,
+			size:    0,
+			mapKeys: make([]string, shardSize),
+			items:   make(map[string]interface{}),
+		}
 	}
 	return m
 }
 
-// GetShard returns shard under given key
-func (m ConcurrentMap) GetShard(key string) *ConcurrentMapShared {
-	return m[uint(fnv32(key))%uint(SHARD_COUNT)]
+// GetShard returns shard under given key.
+func (m ConcurrentMap) GetShard(key string) *ConcurrentMapShard {
+	return m[uint(fnv32(key))%uint(ShardCount)]
 }
 
+// MSet creates a sharded concurrent map out of a regular map.
 func (m ConcurrentMap) MSet(data map[string]interface{}) {
 	for key, value := range data {
-		shard := m.GetShard(key)
-		shard.Lock()
-		shard.items[key] = value
-		shard.Unlock()
+		m.Set(key, value)
 	}
 }
 
-// Sets the given value under the specified key.
+// Set sets the given value under the specified key.
 func (m ConcurrentMap) Set(key string, value interface{}) {
 	// Get map shard.
 	shard := m.GetShard(key)
 	shard.Lock()
+	_, ok := shard.items[key]
 	shard.items[key] = value
+	if !ok {
+		appendKeyToList(key, shard)
+	}
 	shard.Unlock()
 }
 
-// Callback to return new element to be inserted into the map
+// appendKeyToList adds the key to the list, updates the list index and removes last registration if needed.
+// Should be called under shard mutex
+func appendKeyToList(key string, shard *ConcurrentMapShard) {
+	shard.mapKeys[shard.idxAdd] = key
+	shard.idxAdd++
+	shard.idxAdd %= shard.maxSize
+	shard.size++
```

Comment:
instead of using, keeping and maintaining shard.size, you can safely use len(mapKeys)

Reply:
mapKeys always has length maxSize, size is the actual filled data inside mapKeys, once it first reaches capacity, it will also be maxSize.
But I can remove size and always consider mapKeys filled.

Reply:
ok